### DISCUSSION TITLE
Controlling transaction order during parent-child insert

### DISCUSSION THREAD

**Discussion by diogofernandesc** [sqlalchemy/sqlalchemy#8863] [Usage Questions] [answered]
Hi Guys

So I'm running into an issue when inserting an object which populate child objects in the same instantion step.
The parent object has a roles relationship given by:

```
Roles = relationship(
        "UserRole",
        foreign_keys=ExternalId,
        primaryjoin="or_(User.ExternalId == UserRole.ExternalId, "
                    "User.Id == UserRole.Id)",
        uselist=True
        lazy="joined",
        
    )
 ```
 
 When I instantiate the object and insert with:
 ```
user = User(..., Roles=roles)
db.session.add(user)
db.session.commit()
```
What I can see in echo mode is that actually the transaction order during SQL execution means the roles are inserted before the parent (user) object is. Is there a way to invert this such that the User is inserted before the roles in the transaction?

I have seen that `relationship()` has a `post_update` option which allows you to insert both and only link them after via an update query, interestingly I can see this actually changes the order of inserts to be the parent first but I would prefer to avoid having to use the update statement in this scenario if possible.

Thank you in advance!

**Comment by zzzeek** [comment, marked as answer]
guessing on that your "Roles" is attached to User, which means that "foreign_keys=ExternalId" means that User.ExternalID is the "foreign key", would indicate that rows for User are dependent on rows for UserRole.  seems like you want to reverse that.   also use the foreign() annotation:

```py
Roles = relationship(
        "UserRole",
        primaryjoin="or_(User.ExternalId == foreign(UserRole.ExternalId), "
                    "User.Id == foreign(UserRole.Id))",
        uselist=True
        lazy="joined",
        
    )
```

**Reply by diogofernandesc** [reply]
Wow, that makes a lot of sense, thanks a lot Mike, this is simpler than I thought 😆
